src/bin/solang.rs

```diff
@@ -416,33 +431,87 @@ fn compile(matches: &ArgMatches) {
 
     let mut errors = false;
 
-    for filename in matches.get_many::<OsString>("INPUT").unwrap() {
-        match process_file(filename, &mut resolver, target, matches, &mut json, &opt) {
-            Ok(ns) => namespaces.push(ns),
-            Err(_) => {
-                errors = true;
-            }
+    // build a map of requested contract names, and a flag specifying whether it was found or not
+    let mut contract_names: HashSet<&str> = HashSet::new();
+
+    if let Some(values) = matches.get_many::<String>("CONTRACT") {
+        for v in values {
+            contract_names.insert(v.as_str());
         }
     }
 
-    if let Some("ast-dot") = matches.get_one::<String>("EMIT").map(|v| v.as_str()) {
-        exit(0);
+    for filename in matches.get_many::<OsString>("INPUT").unwrap() {
+        // TODO: this could be parallelized using e.g. rayon
+        let ns = process_file(filename, &mut resolver, target, matches, &opt);
+
+        namespaces.push((ns, filename));
     }
 
+    let mut json_contracts = HashMap::new();
+
     let std_json = *matches.get_one("STD-JSON").unwrap();
 
-    if errors {
+    for (ns, _) in &namespaces {
         if std_json {
-            println!("{}", serde_json::to_string(&json).unwrap());
-            exit(0);
+            let mut out = ns.diagnostics_as_json(&resolver);
+            json.errors.append(&mut out);
         } else {
-            eprintln!("error: not all contracts are valid");
-            exit(1);
+            ns.print_diagnostics(&resolver, verbose);
+        }
+
+        if ns.diagnostics.any_errors() {
+            errors = true;
+        }
+    }
+
+    if let Some("ast-dot") = matches.get_one::<String>("EMIT").map(|v| v.as_str()) {
+        exit(0);
+    }
+
+    // ensure we have any contracts
+    if !errors && !namespaces.iter().any(|(ns, _)| !ns.contracts.is_empty()) {
```

Comment:
```suggestion
    if !errors && namespaces.iter().all(|(ns, _)| ns.contracts.is_empty()) {
```

This is written off my head (didn't test), but it would make it much easier to grasp